README.md
# prosemirror

[ [**WEBSITE**](http://prosemirror.net) | [**ISSUES**](https://github.com/prosemirror/prosemirror/issues) | [**FORUM**](https://discuss.prosemirror.net) | [**GITTER**](https://gitter.im/ProseMirror/prosemirror) ]

ProseMirror is a well-behaved rich semantic content editor based on
contentEditable, with support for collaborative editing and custom
document schemas.

The ProseMirror library consists of a number of separate
[modules](https://github.com/prosemirror/). This repository just
serves as a central issue tracker, and holds a script to help easily
check out all the core modules for development.

The [project page](http://prosemirror.net) has more information, a
number of [examples](http://prosemirror.net/examples/) and the
[documentation](http://prosemirror.net/docs/).

This code is released under an
[MIT license](https://github.com/prosemirror/prosemirror/tree/master/LICENSE).
There's a [forum](http://discuss.prosemirror.net) for general
discussion and support requests, and the
[Github bug tracker](https://github.com/prosemirror/prosemirror/issues)
is the place to report issues.

**STOP READING HERE IF YOU'RE SIMPLY _USING_ PROSEMIRROR. YOU CAN
INSTALL THE SEPARATE [NPM
MODULES](https://www.npmjs.com/search?q=prosemirror-) FOR THAT. THE
INSTRUCTIONS BELOW ONLY APPLY WHEN _DEVELOPING_ PROSEMIRROR!**

## Setting up a dev environment

Clone this repository, and make sure you have
[node](https://nodejs.org/en/) and [yarn](https://yarnpkg.com/) (due
to a string of issues with NPM 5, NPM is not currently supported)
installed. Next, from the cloned directory run:

    bin/pm install

This will fetch the submodules, install their dependencies, and build
them.

The `bin/pm` script in this repository provides functionality for
working with the repositories:

 * `bin/pm build` rebuilds all the modules

 * `bin/pm watch` sets up a process that automatically rebuilds the
   modules when they change

 * `bin/pm status` prints the git status of all submodules

 * `bin/pm commit <args>` runs `git commit` with the given arguments
   in all submodules that have pending changes

 * `bin/pm test` runs the (non-browser) tests in all modules

 * `bin/pm push` runs `git push` in all modules.

 * `bin/pm grep <pattern>` greps through the source code for the
   modules for the given pattern

(Functionality for managing releases will be added in the future.)

## Running the demo

To run the demo in `demo/`, do `npm run demo`, and go to
[localhost:8080](http://localhost:8080/). This loads the individual
JavaScript files from the distribution's `dist` directories, and will
only need a refresh when those are changed.

We aim to be an inclusive, welcoming community. To make that explicit,
we have a [code of
conduct](http://contributor-covenant.org/version/1/1/0/) that applies
to communication around the project.


CHANGELOG.md
## [prosemirror-view](http://prosemirror.net/docs/ref/#view) 1.1.1 (2018-03-01)

### Bug fixes

Fixes typo that broke paste.

## [prosemirror-view](http://prosemirror.net/docs/ref/#view) 1.1.0 (2018-02-28)

### Bug fixes

Fixes issue where dragging a draggable node directly below a selected node would move the old selection rather than the target node.

A drop that can't fit the dropped content will no longer dispatch an empty transaction.

### New features

Transactions generated for drop now have a `"uiEvent"` metadata field holding `"drop"`. Paste and cut transactions get that field set to `"paste"` or `"cut"`.

## [prosemirror-transform](http://prosemirror.net/docs/ref/#transform) 1.0.4 (2018-02-23)

### Bug fixes

Fix overeager closing of destination nodes when fitting a slice during replacing.

## [prosemirror-keymap](http://prosemirror.net/docs/ref/#keymap) 1.0.1 (2018-02-23)

### Bug fixes

Upgrade `w3c-keyname` dependency to version 1.1.8 to prevent users getting stuck with a buggy version.

## [prosemirror-transform](http://prosemirror.net/docs/ref/#transform) 1.0.3 (2018-02-23)

### Bug fixes

Fix a problem where slice-placing didn't handle content matches correctly and might generate invalid steps or fail to generate steps though a valid one exists.

Allows adjacent nodes from an inserted slice to be placed in different parent nodes, allowing `Transform.replace` to create fits that weren't previously found.

## [prosemirror-view](http://prosemirror.net/docs/ref/#view) 1.0.11 (2018-02-16)

### Bug fixes

Fix issue where the cursor was visible when a node was selected on recent Chrome versions.

## [prosemirror-gapcursor](http://prosemirror.net/docs/ref/#gapcursor) 1.0.1 (2018-02-16)

### Bug fixes

Prevent issue where clicking on a selectable node near a valid gap cursor position would create a gap cursor rather than select the node.

## [prosemirror-commands](http://prosemirror.net/docs/ref/#commands) 1.0.5 (2018-01-30)

### Bug fixes

Fix crash in [`splitBlock`](http://prosemirror.net/docs/ref/#commands.splitBlock) when `defaultContentType` returns null.

## [prosemirror-transform](http://prosemirror.net/docs/ref/#transform) 1.0.2 (2018-01-24)

### Bug fixes

Fixes a crash in [`replace`](http://prosemirror.net/docs/ref/#transform.Transform.replace).

## [prosemirror-view](http://prosemirror.net/docs/ref/#view) 1.0.10 (2018-01-24)

### Bug fixes

Improve preservation of open and closed nodes in slices taken from the clipboard.

## [prosemirror-state](http://prosemirror.net/docs/ref/#state) 1.1.0 (2018-01-22)

### New features

[`EditorState.toJSON`](http://prosemirror.net/docs/ref/#state.EditorState.toJSON) now accepts a string or number as argument to support the way `JSON.stringify` can call it.

## [prosemirror-commands](http://prosemirror.net/docs/ref/#commands) 1.0.4 (2018-01-18)

### Bug fixes

Pressing delete in front of an [isolating](http://prosemirror.net/docs/ref/#model.NodeSpec.isolating) node no longer opens it.

## [prosemirror-view](http://prosemirror.net/docs/ref/#view) 1.0.9 (2018-01-17)

### Bug fixes

Work around a Chrome cursor motion bug by making sure <br> nodes don't get a contenteditable=false attribute.

## [prosemirror-view](http://prosemirror.net/docs/ref/#view) 1.0.8 (2018-01-09)

### Bug fixes

Fix issue where [`Decoration.map`](http://prosemirror.net/docs/ref/#view.DecorationSet.map) would in some situations with nested nodes incorrectly map decoration positions.

## [prosemirror-view](http://prosemirror.net/docs/ref/#view) 1.0.7 (2018-01-05)

### Bug fixes

Pasting from an external source no longer opens isolating nodes like table cells.

## [prosemirror-model](http://prosemirror.net/docs/ref/#model) 1.1.0 (2018-01-05)

### New features

[`Slice.maxOpen`](http://prosemirror.net/docs/ref/#model.Slice^maxOpen) now has a second argument that can be used to prevent it from opening isolating nodes.

## [prosemirror-view](http://prosemirror.net/docs/ref/#view) 1.0.6 (2017-12-26)

### Bug fixes

[`DecorationSet.remove`](http://prosemirror.net/docs/ref/#view.DecorationSet.remove) now uses a proper deep compare to determine if widgets are the same (it used to compare by identity).

## [prosemirror-commands](http://prosemirror.net/docs/ref/#commands) 1.0.3 (2017-12-19)

### Bug fixes

Fix issue where [`joinBackward`](http://prosemirror.net/docs/ref/#commands.joinBackward) would sometimes create an invalid selection.

## [prosemirror-state](http://prosemirror.net/docs/ref/#state) 1.0.2 (2017-12-13)

### Bug fixes

Fix issue where a character might be selected after overwriting across block nodes.

Make sure `replaceSelectionWith` doesn't needlessly copy unmarked nodes.

## [prosemirror-view](http://prosemirror.net/docs/ref/#view) 1.0.5 (2017-12-05)

### Bug fixes

Fix an issue where deeply nested decorations were mapped incorrectly in corner cases.

## [prosemirror-view](http://prosemirror.net/docs/ref/#view) 1.0.4 (2017-11-27)

### Bug fixes

Fix a corner-case crash during drop.

## [prosemirror-view](http://prosemirror.net/docs/ref/#view) 1.0.3 (2017-11-23)

### Bug fixes

Pressing backspace between two identical characters will no longer generate a transaction that deletes the second one.

## [prosemirror-commands](http://prosemirror.net/docs/ref/#commands) 1.0.2 (2017-11-21)

### Bug fixes

[`splitBlock`](http://prosemirror.net/docs/ref/#commands.splitBlock) no longer crashes when used in a block that's it's parent node's only allowed child.

## [prosemirror-view](http://prosemirror.net/docs/ref/#view) 1.0.2 (2017-11-20)

### Bug fixes

Fix test for whether a node can be selected when arrowing onto it from the right.

Calling [`posAtCoords`](http://prosemirror.net/docs/ref/#view.EditorView.posAtCoords) while a read from the DOM is pending will no longer return a malformed result.

## [prosemirror-view](http://prosemirror.net/docs/ref/#view) 1.0.1 (2017-11-10)

### Bug fixes


src/keys.js
// From CodeMirror, should be factored into its own NPM module

export const names = {
  3: "Enter", 8: "Backspace", 9: "Tab", 13: "Enter", 16: "Shift", 17: "Ctrl", 18: "Alt",
  19: "Pause", 20: "CapsLock", 27: "Esc", 32: "Space", 33: "PageUp", 34: "PageDown", 35: "End",
  36: "Home", 37: "Left", 38: "Up", 39: "Right", 40: "Down", 44: "PrintScrn", 45: "Insert",
  46: "Delete", 59: ";", 61: "=", 91: "Mod", 92: "Mod", 93: "Mod", 107: "=", 109: "-", 127: "Delete",
  173: "-", 186: ";", 187: "=", 188: ",", 189: "-", 190: ".", 191: "/", 192: "`", 219: "[", 220: "\\",
  221: "]", 222: "'", 63232: "Up", 63233: "Down", 63234: "Left", 63235: "Right", 63272: "Delete",
  63273: "Home", 63275: "End", 63276: "PageUp", 63277: "PageDown", 63302: "Insert"
}

// Number keys
for (let i = 0; i < 10; i++) names[i + 48] = names[i + 96] = String(i)
// Alphabetic keys
for (let i = 65; i <= 90; i++) names[i] = String.fromCharCode(i)
// Function keys
for (let i = 1; i <= 12; i++) names[i + 111] = names[i + 63235] = "F" + i

export function keyName(event, noShift) {
  let base = names[event.keyCode], name = base;
  if (name == null || event.altGraphKey) return false

  if (event.altKey && base != "Alt") name = "Alt-" + name
  if (event.ctrlKey && base != "Ctrl") name = "Ctrl-" + name
  if (event.metaKey && base != "Cmd") name = "Cmd-" + name
  if (!noShift && event.shiftKey && base != "Shift") name = "Shift-" + name
  return name
}

export function isModifierKey(value) {
  let name = typeof value == "string" ? value : names[value.keyCode]
  return name == "Ctrl" || name == "Alt" || name == "Shift" || name == "Mod"
}

function normalizeKeyName(name) {
  let parts = name.split(/-(?!$)/), name = parts[parts.length - 1]
  let alt, ctrl, shift, cmd
  for (let i = 0; i < parts.length - 1; i++) {
    let mod = parts[i]
    if (/^(cmd|meta|m)$/i.test(mod)) cmd = true
    else if (/^a(lt)?$/i.test(mod)) alt = true
    else if (/^(c|ctrl|control)$/i.test(mod)) ctrl = true
    else if (/^s(hift)$/i.test(mod)) shift = true
    else throw new Error("Unrecognized modifier name: " + mod)
  }
  if (alt) name = "Alt-" + name
  if (ctrl) name = "Ctrl-" + name
  if (cmd) name = "Cmd-" + name
  if (shift) name = "Shift-" + name
  return name
}

export function normalizeKeymap(keymap) {
  let result = {};
  for (let keyname in keymap) if (keymap.hasOwnProperty(keyname)) {
    let value = keymap[keyname]
    if (value == "...") continue
    if (/^(name|fallthrough|(de|at)tach)$/.test(keyname)) {
      result[keyname] = value
      continue
    }

    let keys = keyname.split(" ").map(normalizeKeyName)
    for (let i = 0; i < keys.length; i++) {
      let val, name
      if (i == keys.length - 1) {
        name = keyname
        val = value
      } else {
        name = keys.slice(0, i + 1).join(" ")
        val = "..."
      }
      let prev = result[name]
      if (!prev) result[name] = val
      else if (prev != val)
        throw new Error("Inconsistent bindings for " + name)
    }
  }
  return result
}

export function lookupKey(key, map, handle, context) {
  let found = map.call ? map.call(null, key, context) : map[key]
  if (found === false) return "nothing"
  if (found === "...") return "multi"
  if (found != null && handle(found)) return "handled"

  if (map.fallthrough) {
    if (Array.isArray(map.fallthrough))
      return lookupKey(key, map.fallthrough, handle, context)
    for (let i = 0; i < map.fallthrough.length; i++) {
      let result = lookupKey(key, map.fallthrough[i], handle, context)
      if (result) return result
    }
  }
}
